akka-cluster-tools/src/main/scala/akka/cluster/pubsub/DistributedPubSubMediator.scala

```diff
@@ -303,28 +384,46 @@ object DistributedPubSubMediator {
       }
     }
 
-    class Topic(val emptyTimeToLive: FiniteDuration, routingLogic: RoutingLogic) extends TopicLike {
+    class Topic(val emptyTimeToLive: FiniteDuration, routingLogic: RoutingLogic, bufferLimit: Int) extends TopicLike with PerGroupingBuffer {
       def business = {
         case msg @ Subscribe(_, Some(group), _) ⇒
           val encGroup = encName(group)
-          context.child(encGroup) match {
-            case Some(g) ⇒ g forward msg
-            case None ⇒
-              val g = context.actorOf(Props(classOf[Group], emptyTimeToLive, routingLogic), name = encGroup)
-              g forward msg
-              context watch g
-              context.parent ! RegisterTopic(g)
+          bufferOr(bufferLimit, mkKey(self.path / encGroup), msg, sender(), msg ⇒ unhandled(msg)) {
+            context.child(encGroup) match {
+              case Some(g) ⇒ g forward msg
+              case None ⇒ newGroupActor(encGroup) forward msg
+            }
           }
           pruneDeadline = None
         case msg @ Unsubscribe(_, Some(group), _) ⇒
-          context.child(encName(group)) match {
-            case Some(g) ⇒ g forward msg
-            case None    ⇒ // no such group here
+          val encGroup = encName(group)
+          bufferOr(bufferLimit, mkKey(self.path / encGroup), msg, sender(), msg ⇒ unhandled(msg)) {
+            context.child(encGroup) match {
+              case Some(g) ⇒ g forward msg
+              case None    ⇒ // no such group here
+            }
           }
         case msg: Subscribed ⇒
           context.parent forward msg
         case msg: Unsubscribed ⇒
           context.parent forward msg
+        case NoMoreSubscribers ⇒
+          val key = mkKey(sender())
+          initializeGrouping(key)
+          sender() ! TerminateRequest
+        case NewSubscriberArrived ⇒
+          val key = mkKey(sender())
+          forwardMessages(key, sender())
```

Comment:
There is one situation where it could happen. Imagine this:

* Topic actor sends `NoMoreSubscribers` to parent
* before `NoMoreSubscribers` arrives to mediator mediator handles `Subscribe` message and forwards it to `Topic`
* `Subscribe` message arrives to `Topic` actor
* `NoMoreSubscribers` arrives to `Mediator`, we start buffering the messages 

now if mediator would only stop `Topic` actor using `context.stop` or send `PoisonPill` subscriber could think that his subscription is registered but we would kill `Topic` actor anyway because we can't do any action when `PoisonPill` is received. 

This scheme works for `cluster-sharding` due to the fact that persistent actors are persistent so even if we terminate it on recreation the state will be restored, in this case we would lose information about subscription that was ACKed just before topic actor was killed.

That is why child actors are not stopped from the parent but additional message is sent to them -  `TerminateRequest`. If child actor sees that it has subscribers/childs it knows that some important message arrived before parent started buffering but after `NoMoreSubscribers` was sent. In that case `NewSubscriberArrived` is sent to parent so it can forward buffered messages to old child actor.